File: client/src/model/PythonFunction.ts
import PythonParameter from "./PythonParameter";
import PythonReturnType from "./PythonReturnType";

export default class PythonFunction {

    readonly name: string;
    readonly decorators: string[];
    // TODO: Implementieren
    readonly parameters: PythonParameter[];
    readonly hasReturnType: boolean;
    // TODO: Implementieren
    readonly returnType: PythonReturnType;
    readonly docstring: string;
    readonly summary: string;
    readonly description: string;

    constructor(
        name: string,
        decorators: string[],
        parameters: PythonParameter[],
        hasReturnType: boolean,
        returnType: PythonReturnType,
        docstring: string,
        summary: string,
        description: string,
    ) {
        this.name = name;
        this.decorators = decorators;
        this.parameters = parameters;
        this.hasReturnType = hasReturnType;
        this.returnType = returnType;
        this.docstring = docstring;
        this.summary = summary;
        this.description = description;
    }
}

File: api-editor/desktop/src/main/kotlin/com/larsreimann/apiEditor/features/ast/model/PythonFunction.kt
package com.larsreimann.apiEditor.features.ast.model

import kotlinx.serialization.Serializable

interface PythonFunction : PythonDeclaration {
    override val id: PythonFunctionId
    val decorators: List<PythonDecorator>
    val documentation: PythonFunctionDocumentation
    val parameters: List<PythonParameter>
    val results: List<PythonResult>

    override fun children() = sequence {
        yieldAll(parameters)
        yieldAll(results)
    }

    fun toMutablePythonFunction() = MutablePythonFunction(
        id = id,
        name = name,
        decorators = decorators,
        documentation = documentation,
        parameters = parameters.map { it.toMutablePythonParameter() },
        results = results.map { it.toMutablePythonResult() },
    )
}

class MutablePythonFunction(
    override val id: PythonFunctionId = PythonFunctionId(""),
    override var name: String = "",
    decorators: List<PythonDecorator> = emptyList(),
    override var documentation: PythonFunctionDocumentation = PythonFunctionDocumentation(),
    parameters: List<MutablePythonParameter> = listOf(),
    results: List<MutablePythonResult> = listOf(),
) : MutablePythonDeclaration(), PythonFunction {

    override val decorators = decorators.toMutableList()
    override val parameters = MutableContainmentList(parameters)
    override val results = MutableContainmentList(results)
}

typealias PythonFunctionId = PythonDeclarationId<PythonFunction>

@Serializable
data class PythonFunctionDocumentation(val description: String = "")
